### PR TITLE
Support '+' in encoded query string when rebuilding the URI to https

### PR DESCRIPTION
**Pull request by jebeaudet**
Fixes https://github.com/spring-cloud/spring-cloud-netflix/issues/1291.

I've already signed the contributor's agreement. Thanks for looking!

### CODE REVIEW

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonUtils.java [hunk 1]
```diff
@@ -67,13 +68,17 @@ public static boolean isSecure(IClientConfig config, ServerIntrospector serverIn
...
-	public static URI updateToHttpsIfNeeded(URI uri, IClientConfig config, ServerIntrospector serverIntrospector, Server server) {
+	public static URI updateToHttpsIfNeeded(URI uri, IClientConfig config, ServerIntrospector serverIntrospector,
+			Server server) {
 		String scheme = uri.getScheme();
 		if (!"https".equals(scheme) && isSecure(config, serverIntrospector, server)) {
-			return UriComponentsBuilder.fromUri(uri).scheme("https").build(true)
-					.toUri();
+			UriComponentsBuilder uriComponentsBuilder = UriComponentsBuilder.fromUri(uri).scheme("https");
+			if (uri.getRawQuery() != null) {
+				// https://jira.spring.io/browse/SPR-10172
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonUtils.java]: Can you also add a short comment about explicitly encoding `+`

File: spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/ribbon/RibbonUtilsTests.java [hunk 2]
```diff
@@ -17,20 +17,23 @@
 
 package org.springframework.cloud.netflix.ribbon;
 
-import com.netflix.client.config.CommonClientConfigKey;
+import static org.hamcrest.Matchers.*;
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/ribbon/RibbonUtilsTests.java]: No `*` imports please.

File: spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/ribbon/RibbonUtilsTests.java [hunk 3]
```diff
@@ -77,6 +80,37 @@ public void nonSecureRibbonPropSecureIntrospector() {
...
+		Assert.assertThat("URI should not have been updated since server is not secured.", original,
+				sameInstance(updated));
+	}
+
+	@Test
+	public void uriIsNotChangedWhenServerIsSecuredAndUriAlreadyInHttps() throws URISyntaxException {
+		URI original = new URI("https://foo");
+		URI updated = updateToHttpsIfNeeded(original, SECURE_CONFIG, SECURE_INTROSPECTOR, SERVER);
+		Assert.assertThat("URI should not have been updated since uri is already in https.", original,
+				sameInstance(updated));
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/ribbon/RibbonUtilsTests.java]: I'd say `equals` is probably ok here.

**Comment by jebeaudet** [comment]
I've made the changes and amended my last commit, hopefully it's fine with you.

**Pull request merged**